src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs

```diff
@@ -56,56 +168,139 @@ public override async Task<int> ExecuteAsync()
             }
         }
 
-        private static async Task CloneRemoteRepoAndDependencies(RemoteFactory remoteFactory, string reposFolder, string repoUri, string commit, bool includeToolset, ILogger logger)
+        private static Local HandleRepoAtSpecificHash(string repoPath, string commit, string masterRepoGitDirPath, ILogger log)
         {
-            IRemote rootRepoRemote = await remoteFactory.GetRemoteAsync(repoUri, logger);
-            IEnumerable<DependencyDetail> rootDependencies = await rootRepoRemote.GetDependenciesAsync(repoUri, commit);
-            rootDependencies = FilterToolsetDependencies(rootDependencies, includeToolset);
+            Local local;
 
-            if (!rootDependencies.Any())
+            if (Directory.Exists(repoPath))
             {
-                string repoPath = GetRepoDirectory(reposFolder, repoUri, commit);
-                if (Directory.Exists(repoPath))
+                log.LogDebug($"Repo path {repoPath} already exists, assuming we cloned already and skipping");
+                local = new Local(log, repoPath);
+            }
+            else
+            {
+                log.LogDebug($"Setting up {repoPath} with .gitdir redirect");
+                Directory.CreateDirectory(repoPath);
+                File.WriteAllText(Path.Combine(repoPath, ".git"), GetGitDirRedirectString(masterRepoGitDirPath));
+                log.LogInformation($"Checking out {commit} in {repoPath}");
+                local = new Local(log, repoPath);
+                local.Checkout(commit, true);
+            }
+
+            return local;
+        }
+
+        private async static Task HandleMasterCopy(RemoteFactory remoteFactory, string repoUrl, string masterGitRepoPath, string masterRepoGitDirPath, ILogger log)
```

Comment:
You may want to extract some methods out of this. The method is large and somewhat hard to follow with all the if/else statements.

Reply:
Refactored this a bit.

---
In reply to: [275044780](https://github.com/dotnet/arcade-services/pull/326#discussion_r275044780) [](ancestors = 275044780)